mobility/choreographer/test/test_zones_keepout.cc

```diff
@@ -0,0 +1,158 @@
+void MResultCallback(ff_util::FreeFlyerActionState::Enum result_code,
+  ff_msgs::MotionResultConstPtr const& result) {
+  EXPECT_EQ(result->response, ff_msgs::MotionResult::VIOLATES_KEEP_OUT);
+  ROS_ERROR("Test Passed");
+  ros::shutdown();
+}
+
+// Called every time feedback is received for the goal
+void MFeedbackCallback(ff_msgs::MotionFeedbackConstPtr const& feedback) {}
+
+
+
+void StateCallback(const ff_msgs::EkfStateConstPtr& state) {
+  if (!client_t_.IsConnected()) return;
+  else if (state->confidence != ff_msgs::EkfState::CONFIDENCE_GOOD) return;
+  else
+    sub_ekf_.shutdown();
+  sleep(1);
+  stable_ = true;
+  return;
+}
+
+// Keepout zone test
+TEST(choreographer_nominal, ZoneBreach) {
+  // Get the node handle
+  ros::NodeHandle nh;
+
+  tf2_ros::TransformListener tfListener(tf_buffer_);
+
+  // Setup MOBILITY action
+  client_t_.SetConnectedTimeout(30.0);
+  client_t_.SetActiveTimeout(30.0);
+  client_t_.SetResponseTimeout(30.0);
+  client_t_.SetFeedbackCallback(std::bind(
+    MFeedbackCallback, std::placeholders::_1));
+  client_t_.SetResultCallback(std::bind(
+    MResultCallback, std::placeholders::_1, std::placeholders::_2));
+  client_t_.Create(&nh, ACTION_MOBILITY_MOTION);
+
+  // Wait for Ekf to start publishing to send the motion goal
+  sub_ekf_ = nh.subscribe(TOPIC_GNC_EKF, 1000, &StateCallback);
+
+  // This waits until the simulation is in a stable status
+  while (!stable_) ros::spinOnce();  // Mobility
+
+  // Configure the planner
+  ff_util::ConfigClient cfg(&nh, NODE_CHOREOGRAPHER);
+
+  cfg.Set<bool>("enable_collision_checking", true);
+  cfg.Set<bool>("enable_validation", true);
+  cfg.Set<bool>("enable_bootstrapping", true);
+  cfg.Set<bool>("enable_immediate", true);
+  cfg.Set<bool>("enable_timesync", false);
+  cfg.Set<bool>("enable_replanning", true);
+  cfg.Set<bool>("enable_faceforward", false);
+  cfg.Set<std::string>("planner", "trapezoidal");
+
+  if (!cfg.Reconfigure()) {
+    std::cout << "Could not reconfigure the choreographer node " << std::endl;
+    EXPECT_EQ(true, true);
+  }
+
+  // Setup a new mobility goal
+  ff_msgs::MotionGoal goal;
+  goal.command = ff_msgs::MotionGoal::MOVE;
+  goal.flight_mode = ff_msgs::MotionGoal::NOMINAL;
+
+  // Pose that breaks the keepout/keepin zones condition
+  geometry_msgs::PoseStamped pose;
+  geometry_msgs::TransformStamped tfs = tf_buffer_.lookupTransform(
+          std::string(FRAME_NAME_WORLD),
+          "body" ,
+          ros::Time(0));
+  pose.header = tfs.header;
+  // pose.header.frame_id = "world";
+
+  pose.pose.position.x = 10.5;
+  pose.pose.position.y = -11.5;
+  pose.pose.position.z = 4.0;
+  pose.pose.orientation.x = 0;
+  pose.pose.orientation.y = 0;
+  pose.pose.orientation.z = 0;
+  pose.pose.orientation.w = 1;
+
+  // Package up and send the move goal
+  goal.states.push_back(pose);
+  // Try and send the goal
+  if (!client_t_.SendGoal(goal)) {
+    std::cout << "Mobility client did not accept goal" << std::endl;
```

Comment:
If this happens, shouldn't we fail the test?

Reply:
ASSERT_EQ(true, false);